File: tests/FSharp.Compiler.ComponentTests/EmittedIL/FixedBindings/PinCSharpByrefStructTypeWithGetPinnableReferenceReturningByref.cs
namespace CsLib
{
    public ref struct RefField<T>
    {
        private readonly ref T _value;

        public T Value
        {
            get => _value;
            set => _value = value;
        }

        public RefField(ref T value)
        {
            this._value = ref value;
        }

        public ref T GetPinnableReference() => ref _value;
    }
}

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/FixedBindings/PinCSharpByrefStructTypeWithGetPinnableReferenceReturningByref.fs
module FixedBindings
open Microsoft.FSharp.NativeInterop
open CsLib

let pinIt (refField: RefField<int>) =
    use ptr = fixed refField
    NativePtr.get ptr 0

[<EntryPoint>]
let main _ =
    let mutable x = 42
    let refToX = new RefField<_>(&x)
    let y = pinIt refToX
    if y <> x then failwith "y did not equal x"
    0